Mathlib/Algebra/Group/Defs.lean

```diff
@@ -620,11 +620,139 @@ the file `Algebra.GroupPower.Basic`. For now, we only register the most basic pr
+  decreasing_by
+    simp_wf
+    exact Nat.div_lt_self (Nat.zero_lt_succ _) (Nat.lt_succ_self _)
+
+/-- Exponentiation by repeated squaring. -/
+-- The unused arguments are essential so that it is possible to prove the `csimp` theorem
+-- `npowRec_eq_npowBinRec`
+@[nolint unusedArguments, to_additive existing]
+def npowBinRec {M : Type*} [Semigroup M] [One M]
+    (_mul_one : ∀ m : M, m * 1 = m) (_one_mul : ∀ m : M, 1 * m = m) (k : ℕ) (m : M) : M :=
+  go k 1 m
+where
+  /-- Auxiliary tail-recursive implementation for `npowBinRec`-/
+  go : ℕ → M → M → M
+  | 0, y, _ => y
+  | (k + 1), y, x =>
+    let k' := (k + 1) >>> 1
+    if k &&& 1 = 1 then
+      go k' y (x * x)
+    else
+      go k' (y * x) (x * x)
+  termination_by go k _ _ => k
+  -- We use this `decreasing_by` block rather than inline `have` statements to avoid
+  -- https://github.com/leanprover/std4/issues/428
+  decreasing_by
+    simp_wf
+    exact Nat.div_lt_self (Nat.zero_lt_succ _) (Nat.lt_succ_self _)
+
+attribute [to_additive existing] npowBinRec.go
+
+@[to_additive]
+theorem npowBinRec.go_spec {M : Type*} [Semigroup M] [One M]
+    (mul_one : ∀ m : M, m * 1 = m) (one_mul : ∀ m : M, 1 * m = m) (k : ℕ) (x y : M) :
+    npowBinRec.go k y x = y * npowRec' mul_one one_mul k x := by
+  induction k using Nat.strongInductionOn generalizing x y with
+  | ind k' ih =>
+    cases k' with
+    | zero => rw [go, npowRec', npowRec, mul_one]
+    | succ k' =>
+      rw [go]
+      dsimp
+      split <;> rename_i w
+      · rw [ih]
+        · congr 1
+          rw [← npowRec'_two_mul]
+          congr 1
+          rw [Nat.shiftRight_eq_div_pow, Nat.pow_one, Nat.mul_div_cancel']
+          simp only [Nat.and_one_is_mod] at w
+          rw [Nat.dvd_iff_mod_eq_zero, Nat.add_mod, w]
+          decide
+        · exact Nat.div_lt_self (Nat.zero_lt_succ _) (Nat.lt_succ_self _)
+      · rw [ih]
+        · rw [mul_assoc y x]
+          rw [← npowRec'_two_mul]
+          congr 1
+          conv =>
+            rhs
+            rw [npowRec', npowRec]
+          congr 2
+          rw [Nat.shiftRight_eq_div_pow, Nat.pow_one]
+          simp only [Nat.and_one_is_mod] at w
+          replace w : k' % 2 = 0 := by
+            cases (Nat.mod_two_eq_zero_or_one k')
+            · assumption
+            · contradiction
+          conv =>
+            lhs
+            rw [← Nat.div_add_mod k' 2]
+          rw [w, Nat.add_zero, Nat.mul_add_div (by decide), show 1 / 2 = 0 by decide, Nat.add_zero,
+            Nat.mul_div_cancel']
+          rwa [Nat.dvd_iff_mod_eq_zero]
+        · exact Nat.div_lt_self (Nat.zero_lt_succ _) (Nat.lt_succ_self _)
+
+@[csimp] theorem nsmulRec_eq_nsmulBinRec : @nsmulRec' = @nsmulBinRec := by
+  funext M _ _ add_zero zero_add k m
+  dsimp [nsmulBinRec]
+  rw [nsmulBinRec.go_spec add_zero, zero_add]
+
+@[csimp] theorem npowRec_eq_npowBinRec : @npowRec' = @npowBinRec := by
+  funext M _ _ mul_one one_mul k m
+  dsimp [npowBinRec]
+  rw [npowBinRec.go_spec mul_one, one_mul]
```

Comment:
To avoid `Group/Defs` growing with things that aren't really the definition of groups, can we move these (and the `BinRec` functions) to `Algebra/Group/FastPower` or something?

Reply:
Another option would be to mark `npowRec'` as `noncomputable` and add a doc comment saying "Import `Algebra.Group.FastPower` to use `npowRec'` inside computable definitions." I believe adding a `@[csimp]` lemma does make noncomputable declarations computable.

Reply:
I would prefer this as a follow up if it is desired.